### DISCUSSION TITLE
High keyboard latency on frame limiter < 8x

### DISCUSSION THREAD

**Discussion by Thesola10** [ppy/osu#23644] [Q&A] [answered]
### Type

Performance

### Bug description

I play osu!lazer on Arch Linux, with `SDL_VIDEODRIVER=wayland`. On both OpenGL and legacy renderers, setting the frame limiter to anything below 8x causes **keyboard** input lag specifically to increase, up to approx. 20ms (from hit error w/o hitsounds). This **does not** affect mouse/tablet input, HID controllers, or MIDI controllers. I tried all of them, and none exhibit the input lag the keyboard has. 8x limiter yields much better latency.

The behavior doesn't appear to be correlated to the actual achieved framerate, as it doesn't really go above 140 on high limiters.

I'm running GNOME 43 by the way, and I always make sure IBus is disabled when playing.

I should note that, while I put the current version in the version area, the issue has been manifesting for a while.

### Version

2023.513.0

### Logs

[runtime.log](https://github.com/ppy/osu/files/11548841/runtime.log)
[performance-update.log](https://github.com/ppy/osu/files/11548842/performance-update.log)
[performance-draw.log](https://github.com/ppy/osu/files/11548843/performance-draw.log)
[performance-audio.log](https://github.com/ppy/osu/files/11548844/performance-audio.log)
[performance.log](https://github.com/ppy/osu/files/11548845/performance.log)
[network.log](https://github.com/ppy/osu/files/11548846/network.log)
[input.log](https://github.com/ppy/osu/files/11548848/input.log)
[database.log](https://github.com/ppy/osu/files/11548849/database.log)

[logs.tar.gz](https://github.com/ppy/osu/files/11548847/logs.tar.gz)

**Comment by peppy** [comment]
If you're measuring keyboard press until audio output, that's not keyboard latency you're measuring. That's full system. You can expect better results with higher frame limiters for the time being as it will increase the number of times a second the update loop is run.

**Reply by Thesola10** [reply]
How does that explain the discrepancy between input methods? Everything lines up in terms of latency except keyboard-to-event. Event-to-video and event-to-audio are short, gamepad-to-event is short, midi-to-event is short, mouse-to-event is short, only the keyboard exhibits this behavior

I was only stating how I got the _number_, of course I didn't stop at key-to-audio.

**Reply by peppy** [reply]
If it's only keyboard, I can only guess that it's something on your system. Maybe resources being stolen somewhere, or CPU or some other hardware entering a higher power state when you increase the limiter, or who knows. We can't really support linux because each user has their system configured in their own way.

**Reply by Thesola10** [reply]
turns out the actual answer was extremely not obvious, and extremely specific to Alder Lake laptops.
[right here](https://github.com/ppy/osu/discussions/23644#discussioncomment-7296341)

**Comment by Thesola10** [comment, marked as answer]
Okay, now I believe I have a scientific answer for this. I was using `taskset` to lock osu! to run on P-cores on my laptop, which includes CPU 0. As I'm running Linux however, CPU 0 is also the boot CPU, which handles all the unsolicited I/O interrupts. Changing the `taskset` command to *avoid* CPU 0 massively improved latency.

Also, I don't want to run osu! on E-cores as this also negatively affects latency, and this is another quirk of Intel's Alder Lake architecture -- E-cores just aren't wired for I/O at all, while P-cores are, and running osu! on all P-cores likely just blocked all I/O because of graphical IRQs, cross-core communication, or whatever. Gamepads worked fine because they're also IRQ-driven, not interrupt-driven like the onboard PS/2 keyboard and trackpad bus on my laptop.

**Reply by Thesola10** [reply]
This also likely won't be reproducible on the newest Intel CPUs as they announced they'd move I/O handling to E-cores to save power, will be fun to find out how this ruins rhythm games
